src/main/scala/is/hail/variant/MatrixTable.scala

```diff
@@ -2065,6 +2065,71 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) {
     })
   }
 
+  def renameFields(oldToNewRows: java.util.HashMap[String, String],
+    oldToNewCols: java.util.HashMap[String, String],
+    oldToNewEntries: java.util.HashMap[String, String],
+    oldToNewGlobals: java.util.HashMap[String, String]): MatrixTable = {
+
+    val fieldMapRows = oldToNewRows.asScala
+    assert(fieldMapRows.keys.forall(k => matrixType.rowType.fieldNames.contains(k)),
+      s"[${fieldMapRows.keys.mkString(", ")}], expected [${ matrixType.rowType.fieldNames.mkString(", ") }]")
+
+    val fieldMapCols = oldToNewCols.asScala
+    assert(fieldMapCols.keys.forall(k => matrixType.colType.fieldNames.contains(k)),
+      s"[${fieldMapCols.keys.mkString(", ")}], expected [${ matrixType.colType.fieldNames.mkString(", ") }]")
+
+    val fieldMapEntries = oldToNewEntries.asScala
+    assert(fieldMapEntries.keys.forall(k => matrixType.entryType.fieldNames.contains(k)),
+      s"[${fieldMapEntries.keys.mkString(", ")}], expected [${ matrixType.entryType.fieldNames.mkString(", ") }]")
+
+    val fieldMapGlobals = oldToNewGlobals.asScala
+    assert(fieldMapGlobals.keys.forall(k => matrixType.globalType.fieldNames.contains(k)),
+      s"[${fieldMapGlobals.keys.mkString(", ")}], expected [${ matrixType.globalType.fieldNames.mkString(", ") }]")
+
+    val (newColKey, newColType) = if (fieldMapCols.isEmpty) (colKey, colType) else {
+      val newFieldNames = colType.fieldNames.map { n => fieldMapCols.getOrElse(n, n) }
+      val newKey = colKey.map { f => fieldMapCols.getOrElse(f, f) }
+      (newKey, TStruct(newFieldNames.zip(colType.fieldType): _*))
+    }
+
+    val newEntryType = if (fieldMapEntries.isEmpty) entryType else {
+      val newFieldNames = entryType.fieldNames.map { n => fieldMapEntries.getOrElse(n, n) }
+      TStruct(newFieldNames.zip(entryType.fieldType): _*)
+    }
+
+    val (pk, newRowKey, newRVRowType) = {
+      val newPK = rowPartitionKey.map { f => fieldMapRows.getOrElse(f, f) }
+      val newKey = rowKey.map { f => fieldMapRows.getOrElse(f, f) }
+      val newRVRowType = TStruct(rvRowType.fields.map { f =>
+          f.name match {
+            case x@MatrixType.entriesIdentifier => (x, TArray(newEntryType))
```

Comment:
will this array ever be required? if so, this could be a bug.

Reply:
I think safer to not special case the entries identifier (we assume the user never inputs that weird long string anyway)

Reply:
I'm constructing rvrowtype here instead of rowtype since the location of the entries within the struct isn't guaranteed. Special casing here because it's the only case in which the underlying type actually changes. I guess I could construct the rowtype, and then insert the entries field at the correct spot, but I don't think that's necessarily neater.

Reply:
d'oh, totally missed the newEntryType.

Reply:
I'll do a check on the requiredness of the array though.

Reply:
actually, requiredness could be a problem in other places too, right?

Reply:
I'll do a check when I construct the structs, too. The others should be fine because I'm copying the types.